depthai_ros_driver/include/depthai_ros_driver/impl/ros_agnostic.hpp

```diff
@@ -0,0 +1,249 @@
+#define DEPTHAI_ROS_DRIVER__IMPL__ROS_AGNOSTIC_HPP
+
+#include <depthai_ros_driver/ros_agnostic.hpp>
+
+namespace rr {
+namespace ros_agnostic {
+
+/*
+ * Here we will keep all details implemntaion of within ros_agnostic ns. This
+ * includes: Publisher, Subscription, Server, Timer...
+ */
+
+//==============================================================================
+class Publisher
+{
+public:
+  /// @brief Publish ros msg
+  template<typename Msg>
+  void publish(Msg& msg)
+  {
+    #if defined(USE_ROS2)
+    // move the ptr and subscriber check here
+    auto pub = std::static_pointer_cast<rclcpp::Publisher<Msg>>(_ptr);
+    if (pub && pub->get_subscription_count() > 0)
+      pub->publish(msg);
+    #else
+    if (_pub.getNumSubscribers() > 0);
+    _pub.publish(msg);
+    #endif
+  }
+
+private:
+  /// @brief Create ros publisher
+  template<class Msg>
+  void create_publisher(
+    ROSNodeHandle node_handle,
+    const std::string& topic_name,
+    const uint32_t queue_size)
+  {
+    #if defined(USE_ROS2)
+    // using qos: KeepLast(QueueSize)
+    _ptr = node_handle->create_publisher<Msg>(topic_name, queue_size);
+    #else
+    _pub = node_handle->advertise<Msg>(topic_name, queue_size);
+    #endif
+  }
+
+  #if defined(USE_ROS2)
+  std::shared_ptr<void> _ptr;
+  #else
+  ros::Publisher _pub;
+  #endif
+
+  friend class NodeInterface;
+};
+
+//==============================================================================
+class Subscription
+{
+private:
+  /// @brief Create ros subscription
+  template<class Msg, typename Callback>
+  void create_subscription(
+    ROSNodeHandle node_handle,
+    const std::string& topic_name,
+    const uint32_t queue_size,
+    const Callback& callback)
+  {
+    #if defined(USE_ROS2)
+    // using qos: KeepLast(QueueSize)
+    _ptr = node_handle->create_subscription<Msg>(topic_name, queue_size,
+        callback);
+    #else
+    _sub = node_handle->subscribe<Msg>(topic_name, queue_size, callback);
+    #endif
+  }
+
+  #if defined(USE_ROS2)
+  std::shared_ptr<void> _ptr;
+  #else
+  ros::Subscriber _sub;
+  #endif
+
+  friend class NodeInterface;
+};
+
+//==============================================================================
+class Service
+{
+private:
+  /// @brief Create ros service
+  template<class Msg, typename Callback>
+  void create_service(
+    ROSNodeHandle node_handle,
+    const std::string& srv_name,
+    const Callback& callback)
+  {
+    #if defined(USE_ROS2)
+    _ptr = node_handle->create_service<Msg>(srv_name, callback);
+    #else
+    using RequestMsg = typename Msg::Request;
+    using ResponseMsg = typename Msg::Response;
+    _srv = node_handle->advertiseService<RequestMsg, ResponseMsg>(
+      srv_name,
+      [callback = std::move(callback)](const RequestMsg& req, ResponseMsg& res)
+      {
+        // internally convert ref to ptr
+        auto res_ptr = std::make_shared<ResponseMsg>();
+        callback(std::make_shared<RequestMsg>(req), res_ptr);
```

Comment:
Are these make_shared added so that the interface between ROS1 and ROS2 stays the same?

Reply:
If so, please add a comment